### PR TITLE
Compare treatment phenotype predictions

### PR DESCRIPTION
**Pull request by MattsonCam**
In this pr, I refactored the treatment comparisons between phenotypic class probabilities. Additionally, I added only compare treatments to negative controls within the same plate, and don't compare wells with cell counts below a supplied threshold. Please let me know if you have any questions or suggestions.

### CODE REVIEW

File: 3.analyze_data/utils/significance_testing.py [hunk 1]
```diff
@@ -44,7 +43,35 @@ def store_comparisons(_comp_functions, _treatments, _dmso_probs, _treatment_prob
...
+    ----------
+    _df: pandas Dataframe
+        The dataframe which contains the well, which will be potentially removed
+
+    _cutoff: Integer
+        The number of cells a well must have to remain in the dataframe
+
+    Returns
+    -------
+    _df: pandas dataframe
```
Comment (MikeLippincott) [inline 3.analyze_data/utils/significance_testing.py]:
```suggestion
    _df: pandas.DataFrame
```

File: 3.analyze_data/utils/significance_testing.py [hunk 2]
```diff
@@ -44,7 +43,35 @@ def store_comparisons(_comp_functions, _treatments, _dmso_probs, _treatment_prob
 
     return _treatments
 
-def get_treatment_comparison(_comp_functions, _treatment_paths, _probadf):
+def filter_wells_by_cell_count(_df, _cutoff):
+    """
+    Parameters
+    ----------
+    _df: pandas Dataframe
```
Comment (MikeLippincott) [inline 3.analyze_data/utils/significance_testing.py]:
```suggestion
    _df: pandas.DataFrame
```

File: 3.analyze_data/utils/significance_testing.py (hunk 1 above)
Comment (MikeLippincott) [reply 3.analyze_data/utils/significance_testing.py]: This is to denote that the `DataFrame` object is a part of the `pandas` package. Consider adding but doesn't change overall function

Comment (MikeLippincott) [reply 3.analyze_data/utils/significance_testing.py]: Consider applying to the rest of the occurrences as well.

File: 3.analyze_data/utils/significance_testing.py [hunk 3]
```diff
@@ -44,7 +43,35 @@ def store_comparisons(_comp_functions, _treatments, _dmso_probs, _treatment_prob
...
+    -------
+    _df: pandas dataframe
+        The filtered dataframe with only wells that have more cells than the cutoff
+    """
+
+    # Find the number of cells in each well
+    well_counts = _df["Metadata_Well"].value_counts()
+
+    # Find the wells that contain more cells than the cutoff
+    wells_kept = well_counts[well_counts > _cutoff].index
```
Comment (MikeLippincott) [inline 3.analyze_data/utils/significance_testing.py:66]:
```suggestion
    wells_kept_index = well_counts[well_counts > _cutoff].index
```
Consider changing this var name to better reflect what the var represents

File: 3.analyze_data/utils/significance_testing.py [hunk 4]
```diff
@@ -44,7 +43,35 @@ def store_comparisons(_comp_functions, _treatments, _dmso_probs, _treatment_prob
...
+
+    # Keep only the cells that contain more cells than the cutoff
+    _df = _df.loc[_df["Metadata_Well"].isin(wells_kept)]
+
+    return _df
+
+
+def get_treatment_comparison(_comp_functions, _treatment_paths, _probadf, _barcode_platemapdf, _control_cutoff = 50, _treat_cutoff = 50):
```
Comment (MikeLippincott) [inline 3.analyze_data/utils/significance_testing.py:74]: This a really well written function and the docstrings are gorgeous!

File: 3.analyze_data/utils/significance_testing.py [hunk 5]
```diff
@@ -56,12 +83,21 @@ def get_treatment_comparison(_comp_functions, _treatment_paths, _probadf):
...
+    _probadf: pandas Dataframe
         The predicted probabilities and associated metadata for each cell
 
+    _barcode_platemapdf: pandas Dataframe
+        Maps the plate to the treatment type
+
+    _control_cutoff: Integer
+        The minimum number of cells required for a negative control well to be included in the comparison
+
+    _treat_cutoff: Integer
```
Comment (MikeLippincott) [inline 3.analyze_data/utils/significance_testing.py:98]: Also, very descriptive too. This is great documentation!

File: 3.analyze_data/utils/significance_testing.py [hunk 6]
```diff
@@ -70,40 +106,70 @@ def get_treatment_comparison(_comp_functions, _treatment_paths, _probadf):
...
-                # Find the treatment probabilities that correspond to each well, and therefore each treatment
-                treat_probas = filtered_probadf.loc[filtered_probadf["Metadata_Well"] == row["well_position"]]
+            # Compute the number of cells for each group
+            treat_cell_count = len(treatdf)
+            negcon_cell_count = len(negdf)
+            min_cell_count = min(treat_cell_count, negcon_cell_count)
 
-                num_cells = len(treat_probas)
+            # If there are no probability values that match the given well for some reason analyze the next treatment
+            if (min_cell_count > 0):
```
Comment (MikeLippincott) [inline 3.analyze_data/utils/significance_testing.py:157]: Nice filtering for cell counts. Ths is a clever way to go through the wells.

File: environment.yml [hunk 7]
```diff
@@ -3,7 +3,7 @@ channels:
   - conda-forge
 dependencies:
   - conda-forge::python=3.9
-  - pandas
+  - pandas<2.0
```
Comment (MikeLippincott) [inline environment.yml:6]: Good dependency management here with the `<2.0` for pandas

File: 3.analyze_data/nbconverted/analyze_predicted_probabilities.py [hunk 8]
```diff
@@ -138,26 +167,29 @@ def perform_dunn_median(_dmso_probs, _treatment_probs):
 
 # ## Defining tests and aggregation metric names
 
-# In[7]:
+# In[8]:
 
 
-comp_functions = {"dunn_test":
+comp_functions = {"dunn":
```
Comment (MikeLippincott) [inline 3.analyze_data/nbconverted/analyze_predicted_probabilities.py:175]: consider adding a comment here to explain  the dict

**Review by MikeLippincott — APPROVED** [review]
Nice, clean PR. I think you did an amazing job with your comments and inline documentation! This PR is ready to merge after addressing some small comments. I have 1 suggestion that I think should probably be a separate PR. I think this repo is in need of an additional README file that explains the treatments, cells, plates etc. The code are pretty abstract at the moment and it is fairly hard to understand how the analysis ties into the dataset and what it all means. This is entirely up to you and you might be able to link another REPO's README in the main README on this repo to address this.

File: 3.analyze_data/utils/significance_testing.py (hunk 3 above)
Comment (MattsonCam) [reply 3.analyze_data/utils/significance_testing.py:66]: The index in this case is just how the wells are stored

File: 3.analyze_data/utils/significance_testing.py (hunk 1 above)
Comment (MattsonCam) [reply 3.analyze_data/utils/significance_testing.py]: Changed! Also did this for pandas series

File: 3.analyze_data/nbconverted/analyze_predicted_probabilities.py (hunk 8 above)
Comment (MattsonCam) [reply 3.analyze_data/nbconverted/analyze_predicted_probabilities.py:175]: Added!

**Comment by MattsonCam** [comment]
Thanks Mike, merging now!

**Pull request merged**